rascal-lsp/src/main/rascal/util/LanguageServer.rsc

```diff
@@ -268,6 +272,10 @@ data LanguageService
     | references    (set[loc] (Focus _focus) referencesService)
     | implementation(set[loc] (Focus _focus) implementationService)
     | codeAction    (list[CodeAction] (Focus _focus) codeActionService)
+    | prepareRename (loc (Focus _focus) prepareRenameService)
+    | rename        (tuple[list[DocumentEdit], set[Message]] (Focus _focus, str newName) renameService)
+    | willRenameFiles(tuple[list[DocumentEdit], set[Message]] (map[loc old, loc new] fileRenames) willRenameFilesService)
+    | didRenameFiles(void(map[loc old, loc new] fileRenames) didRenameFilesService)
```

Comment:
The `map[loc, loc]` that contains the renamed files could also be a `rel[loc, loc]`. That is closer to the LSP spec, which uses a `FileRename[]` here (`FileRename` is basically a named pair). However, the `rel` is a bit confusing; can we have multiple occurrences of original file names, renamed to different target names?

Reply:
I think if it's a list in LSP, it might map better onto an `lrel` to keep the order.